README.md

```diff
@@ -261,6 +381,161 @@ class MyChain extends SetupChain {
+    return opts.name || randomName()
+  }
+
+  greet: async (opts) => {
+    const defaults = {
+      greeting: 'hello'
+    , names: []
+    }
+
+    const config = this.lookup({
+      ...defaults
+    , ...opts
+    })
+
+    return config.names.map((name) => {
+      return `${config.greeting} ${name}`
+    })
+  }
+}
+
+class MyChain extends SetupChain {
+  constructor(state) {
+    super(state, actions)
+  }
+}
+```
+
+The result of previous function calls in the chain can be passed as arguments into another.
+The values of previous results can be accessed using the `#` symbol followed by the name
+of the key. Simple Object path notation is supported
+
+```javascript
+const state = await new MyChain()
+  .name({name: 'greg'}, 'name_1')
+  .name({name: 'fred'}, 'name_2')
+  .greet({
+    greeting: 'Goodbye'
+  , names: ['#name_1', '#name_2']
+  }, 'greeting_1')
+  .execute()
+
+console.log(state)
+
+{
+  name_1: 'greg'
+  name_2: 'fred'
+, greeting_1: ['Goodbye greg', 'Goodbye fred']
+}
+
+```
+
+## Helper Functions via `lookup`
+
+The `lookup` function has the ability to execute functions when simple object path lookups aren't sufficient.
+The syntax for function execution is as follows. Simple arguments can be passed.
+If arguments are passed, numeric and boolean values will be casted to the
+appropriate type. Everything else will be handled as a string. String arguments
+must be quoted with either single or double quotes if the string contains a comma.
+Functions may also be used as arguments, but must use the more conventional call `()`
+syntax to ensure arguments are passed appropriately.
+
+```
+!<name>:arg,arg,arg
+!<name>:"one,two",three
+!<name>("one,two", !random:1, !foo("bar", "baz"), #nested.key)
+```
+
+### Included Helper Functions
+
+There is a few helper functions that we felt were valuable enough to include in
+the base class.
+
+### `random(<Number>)`
+Generates a random HEX string. It accepts an optional single argument that specifies
+the number of random bytes to generate. This can be used when generating
+unique ids or names to be used in database records, for example. Combined with
+`!template`, this is a useful helper function.
+
+**Parameters**
+
+* **bytes** [Number][]: The number of bytes in the result
+
+**Returns:** [String][]
+
+```javascript
+chain.lookup('!random:2') // 3830
+chain.lookup('!random:10') // eddbdf576eac2ded313d
+```
+
+### `template(<String>)`
+Returns a string with replacement patterns from the chain. Templates are rendered in the
+same sequence as other operations on the chain. Only the data from actions prior to the
+template function will be available for replacement.
+
+**Parameters**
+
+* **template** [String][]: The string template to parse
+
+**Returns:** [String][]
+
+Templates supports a basic bracket syntax for replacements:
+
+```javascript
+chain.set('name', 'World')
+chain.set('foo', {bar: "baz"})
+chain.lookup('!template:"Hello {{#name}}"') // Hello World
```

Comment:
The examples here are also a bit confusing and out of context.  `set` is an action, so it requires `execute` to technically be correct here.  Also, `chain.lookup` looks like an action, but it's not.  I think `lookup` needs to be an action to resolve this.